File: mdix-php/tests/BuilderTest.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix\Tests;

use MidManStudio\Mdix\ErrorKind;
use MidManStudio\Mdix\MdixBuilder;
use MidManStudio\Mdix\MdixError;
use PHPUnit\Framework\TestCase;

final class BuilderTest extends TestCase
{
    // ── Lifecycle ─────────────────────────────────────────────────────────────

    public function testNewBuilderIsNotNull(): void
    {
        $b = new MdixBuilder();
        self::assertNotNull($b);
        $b->close();
    }

    public function testCloseCalledTwiceDoesNotThrow(): void
    {
        $b = new MdixBuilder();
        $b->close();
        $b->close();
        self::assertTrue(true);
    }

    public function testSetStringAfterCloseThrows(): void
    {
        $b = new MdixBuilder();
        $b->close();

        $this->expectException(MdixError::class);
        $b->setString('x', 'v');
    }

    public function testSetStringAfterCloseErrorKind(): void
    {
        $b = new MdixBuilder();
        $b->close();

        try {
            $b->setString('x', 'v');
            self::fail('Expected MdixError');
        } catch (MdixError $e) {
            self::assertSame(ErrorKind::Closed, $e->kind);
        }
    }

    // ── Set / has key ─────────────────────────────────────────────────────────

    public function testSetStringHasKey(): void
    {
        $b = new MdixBuilder();
        $b->setString('app.name', 'DixScript');
        self::assertTrue($b->hasKey('app.name'));
        $b->close();
    }

    public function testSetIntHasKey(): void
    {
        $b = new MdixBuilder();
        $b->setInt('port', 8080);
        self::assertTrue($b->hasKey('port'));
        $b->close();
    }

    public function testSetFloatHasKey(): void
    {
        $b = new MdixBuilder();
        $b->setFloat('rate', 1.5);
        self::assertTrue($b->hasKey('rate'));
        $b->close();
    }

    public function testSetDoubleHasKey(): void
    {
        $b = new MdixBuilder();
        $b->setDouble('pi', 3.14159);
        self::assertTrue($b->hasKey('pi'));
        $b->close();
    }

    public function testSetBoolHasKey(): void
    {
        $b = new MdixBuilder();
        $b->setBool('debug', true);
        self::assertTrue($b->hasKey('debug'));
        $b->close();
    }

    // ── Fluent chaining ───────────────────────────────────────────────────────

    public function testFluentChainWorks(): void
    {
        $b = new MdixBuilder();
        $b->setString('a', 'hello')
          ->setInt('b', 42)
          ->setBool('c', false);
        self::assertTrue($b->hasKey('a'));
        self::assertTrue($b->hasKey('b'));
        self::assertTrue($b->hasKey('c'));
        $b->close();
    }

    // ── Remove ────────────────────────────────────────────────────────────────

    public function testRemoveExistingKeyReturnsTrue(): void
    {
        $b = new MdixBuilder();
        $b->setInt('x', 1);
        self::assertTrue($b->remove('x'));
        self::assertFalse($b->hasKey('x'));
        $b->close();
    }

    public function testRemoveMissingKeyReturnsFalse(): void
    {
        $b = new MdixBuilder();
        self::assertFalse($b->remove('nope'));
        $b->close();
    }

    // ── Clear ─────────────────────────────────────────────────────────────────

    public function testClearRemovesAllKeys(): void
    {
        $b = new MdixBuilder();
        $b->setString('a', '1')->setInt('b', 2)->setBool('c', true);
        $b->clear();
        self::assertFalse($b->hasKey('a'));
        self::assertFalse($b->hasKey('b'));
        self::assertFalse($b->hasKey('c'));
        $b->close();
    }

    // ── toMdixString / toDatabase ─────────────────────────────────────────────

    public function testToMdixStringNonEmpty(): void
    {
        $b = new MdixBuilder();
        $b->setString('name', 'test')->setInt('val', 99);
        self::assertNotEmpty($b->toMdixString());
        $b->close();
    }

    public function testToDatabaseValuesReadable(): void
    {
        $b = new MdixBuilder();
        $b->setString('greet', 'hello')
          ->setInt('num', 7)
          ->setBool('flag', true);

        $db = $b->toDatabase();
        try {
            self::assertSame('hello', $db->getString('greet'));
            self::assertSame(7, $db->getInt('num'));
            self::assertTrue($db->getBool('flag'));
        } finally {
            $db->close();
            $b->close();
        }
    }

    public function testToDatabaseMultipleTypes(): void
    {
        $b = new MdixBuilder();
        $b->setString('s', 'world')
          ->setInt('i', 42)
          ->setFloat('f', 1.5)
          ->setDouble('d', 3.14)
          ->setBool('b', false);

        $db = $b->toDatabase();
        try {
            self::assertSame('world', $db->getString('s'));
            self::assertSame(42, $db->getInt('i'));
            self::assertEqualsWithDelta(1.5, $db->getFloat('f'), 0.001);
            self::assertEqualsWithDelta(3.14, $db->getDouble('d'), 0.001);
            self::assertFalse($db->getBool('b'));
        } finally {
            $db->close();
            $b->close();
        }
    }

    // ── saveToFile ────────────────────────────────────────────────────────────

    public function testSaveToFileWritesFile(): void
    {
        $b    = new MdixBuilder();
        $path = \sys_get_temp_dir() . '/mdix_test_' . \uniqid() . '.mdix';

        try {
            $b->setString('saved', 'yes');
            $b->saveToFile($path);
            self::assertFileExists($path);
        } finally {
            if (\file_exists($path)) {
                \unlink($path);
            }
            $b->close();
        }
    }

    public function testSaveToFileLoadBack(): void
    {
        $b    = new MdixBuilder();
        $path = \sys_get_temp_dir() . '/mdix_test_' . \uniqid() . '.mdix';

        try {
            $b->setInt('answer', 42);
            $b->saveToFile($path);
            $b->close();

            $db = \MidManStudio\Mdix\MdixDatabase::load($path);
            try {
                self::assertSame(42, $db->getInt('answer'));
            } finally {
                $db->close();
            }
        } finally {
            if (\file_exists($path)) {
                \unlink($path);
            }
        }
    }

    // ── null / empty path guards ──────────────────────────────────────────────

    public function testSetStringEmptyPathThrows(): void
    {
        $b = new MdixBuilder();
        try {
            $this->expectException(MdixError::class);
            $b->setString('', 'v');
        } finally {
            $b->close();
        }
    }

    public function testSetIntEmptyPathThrows(): void
    {
        $b = new MdixBuilder();
        try {
            $this->expectException(MdixError::class);
            $b->setInt('', 1);
        } finally {
            $b->close();
        }
    }

    // ── tryToDatabase ─────────────────────────────────────────────────────────

    public function testTryToDatabaseSuccess(): void
    {
        $b      = new MdixBuilder();
        $result = $b->setInt('port', 8080)->tryToDatabase();
        self::assertTrue($result->isSuccess());
        $result->getValue()->close();
        $b->close();
    }

    // ── date / timestamp helpers ──────────────────────────────────────────────

    public function testSetDateStoresString(): void
    {
        $b    = new MdixBuilder();
        $date = new \DateTimeImmutable('2025-06-15');
        $b->setDate('release', $date);
        self::assertTrue($b->hasKey('release'));
        $b->close();
    }

    public function testSetTimestampStoresString(): void
    {
        $b  = new MdixBuilder();
        $ts = new \DateTimeImmutable('2025-06-15T10:30:00Z');
        $b->setTimestamp('created_at', $ts);
        self::assertTrue($b->hasKey('created_at'));
        $b->close();
    }
}

File: mdix-java/src/main/test/java/com/midmanstudio/dixscript/BuilderTest.java
package com.midmanstudio.dixscript;

import org.junit.jupiter.api.*;
import static org.assertj.core.api.Assertions.*;

class BuilderTest {

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    @Test void newBuilder_isNotNull() {
        try (Builder b = DixScript.newBuilder()) {
            assertThat(b).isNotNull();
        }
    }

    @Test void close_calledTwice_doesNotThrow() {
        Builder b = DixScript.newBuilder();
        b.close();
        assertThatCode(b::close).doesNotThrowAnyException();
    }

    @Test void setString_afterClose_throws() {
        Builder b = DixScript.newBuilder();
        b.close();
        assertThatThrownBy(() -> b.setString("x", "v"))
            .isInstanceOf(MdixException.class)
            .extracting(e -> ((MdixException) e).getKind())
            .isEqualTo(MdixException.Kind.CLOSED);
    }

    // ── Set / get round-trip ──────────────────────────────────────────────────

    @Test void setString_getBack_roundTrips() {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("app.name", "DixScript");
            assertThat(b.hasKey("app.name")).isTrue();
        }
    }

    @Test void setInt_hasKey_true() {
        try (Builder b = DixScript.newBuilder()) {
            b.setInt("port", 8080);
            assertThat(b.hasKey("port")).isTrue();
        }
    }

    @Test void setFloat_hasKey_true() {
        try (Builder b = DixScript.newBuilder()) {
            b.setFloat("rate", 1.5f);
            assertThat(b.hasKey("rate")).isTrue();
        }
    }

    @Test void setDouble_hasKey_true() {
        try (Builder b = DixScript.newBuilder()) {
            b.setDouble("pi", 3.14159);
            assertThat(b.hasKey("pi")).isTrue();
        }
    }

    @Test void setBool_hasKey_true() {
        try (Builder b = DixScript.newBuilder()) {
            b.setBool("debug", true);
            assertThat(b.hasKey("debug")).isTrue();
        }
    }

    // ── Fluent chaining ───────────────────────────────────────────────────────

    @Test void fluent_chain_works() {
        try (Builder b = DixScript.newBuilder()) {
            assertThatCode(() ->
                b.setString("a", "hello")
                 .setInt("b", 42)
                 .setBool("c", false)
            ).doesNotThrowAnyException();
        }
    }

    // ── Remove ────────────────────────────────────────────────────────────────

    @Test void remove_existingKey_returnsTrue() {
        try (Builder b = DixScript.newBuilder()) {
            b.setInt("x", 1);
            assertThat(b.remove("x")).isTrue();
            assertThat(b.hasKey("x")).isFalse();
        }
    }

    @Test void remove_missingKey_returnsFalse() {
        try (Builder b = DixScript.newBuilder()) {
            assertThat(b.remove("nope")).isFalse();
        }
    }

    // ── Clear ─────────────────────────────────────────────────────────────────

    @Test void clear_removesAllKeys() {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("a", "1").setInt("b", 2).setBool("c", true);
            b.clear();
            assertThat(b.hasKey("a")).isFalse();
            assertThat(b.hasKey("b")).isFalse();
        }
    }

    // ── toString / toDatabase ─────────────────────────────────────────────────

    @Test void toMdixString_nonEmpty() {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("name", "test").setInt("val", 99);
            String s = b.toMdixString();
            assertThat(s).isNotBlank();
        }
    }

    @Test void toDatabase_valuesReadable() {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("greet", "hello");
            b.setInt("num", 7);
            b.setBool("flag", true);

            try (Database db = b.toDatabase()) {
                assertThat(db.getString("greet")).isEqualTo("hello");
                assertThat(db.getInt("num")).isEqualTo(7);
                assertThat(db.getBool("flag")).isTrue();
            }
        }
    }

    @Test void toDatabase_multipleTypes_allReadable() {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("s", "world")
             .setInt("i", 42)
             .setFloat("f", 1.5f)
             .setDouble("d", 3.14)
             .setBool("b", false);

            try (Database db = b.toDatabase()) {
                assertThat(db.getString("s")).isEqualTo("world");
                assertThat(db.getInt("i")).isEqualTo(42);
                assertThat(db.getFloat("f")).isCloseTo(1.5f, within(0.001f));
                assertThat(db.getDouble("d")).isCloseTo(3.14, within(0.001));
                assertThat(db.getBool("b")).isFalse();
            }
        }
    }

    // ── saveToFile ────────────────────────────────────────────────────────────

    @Test void saveToFile_writesFile(@TempDir java.nio.file.Path tmp) {
        try (Builder b = DixScript.newBuilder()) {
            b.setString("saved", "yes");
            String path = tmp.resolve("out.mdix").toString();
            assertThatCode(() -> b.saveToFile(path)).doesNotThrowAnyException();
            assertThat(new java.io.File(path)).exists();
        }
    }

    @Test void saveToFile_loadBack_valuesPresent(@TempDir java.nio.file.Path tmp) {
        String path = tmp.resolve("roundtrip.mdix").toString();
        try (Builder b = DixScript.newBuilder()) {
            b.setInt("answer", 42);
            b.saveToFile(path);
        }
        try (Database db = DixScript.load(path)) {
            assertThat(db.getInt("answer")).isEqualTo(42);
        }
    }

    // ── Null / empty path guards ──────────────────────────────────────────────

    @Test void setString_nullPath_throws() {
        try (Builder b = DixScript.newBuilder()) {
            assertThatThrownBy(() -> b.setString(null, "v"))
                .isInstanceOf(MdixException.class)
                .extracting(e -> ((MdixException) e).getKind())
                .isEqualTo(MdixException.Kind.INVALID_PATH);
        }
    }

    @Test void setString_emptyPath_throws() {
        try (Builder b = DixScript.newBuilder()) {
            assertThatThrownBy(() -> b.setString("", "v"))
                .isInstanceOf(MdixException.class);
        }
    }
          }
